File: internal/services/upgrade_service.go
package services

import (
	"fmt"
	"sort"

	"github.com/oberones/star-smuggler-go/internal/domain"
)

type UpgradePurchaseResult struct {
	Succeeded bool
	Message   string
}

type UpgradeService struct{}

func SuccessfulUpgradePurchase(message string) UpgradePurchaseResult {
	return UpgradePurchaseResult{
		Succeeded: true,
		Message:   message,
	}
}

func FailedUpgradePurchase(message string) UpgradePurchaseResult {
	return UpgradePurchaseResult{
		Succeeded: false,
		Message:   message,
	}
}

func (s UpgradeService) AvailableUpgrades(run domain.RunState, snapshot domain.DataSnapshot, factions FactionService) []domain.ShipUpgradeDefinition {
	available := make([]domain.ShipUpgradeDefinition, 0)
	for _, upgrade := range snapshot.Upgrades {
		if run.Progression.HasUpgrade(upgrade.ID) {
			continue
		}
		if upgrade.RequiredFactionID != "" &&
			!factions.MeetsMinimumStanding(run, snapshot, upgrade.RequiredFactionID, upgrade.MinimumStanding) {
			continue
		}
		available = append(available, upgrade)
	}

	sort.SliceStable(available, func(i, j int) bool {
		if available[i].CostCredits == available[j].CostCredits {
			return available[i].Name < available[j].Name
		}
		return available[i].CostCredits < available[j].CostCredits
	})

	return available
}

func (s UpgradeService) PurchaseUpgrade(run *domain.RunState, snapshot domain.DataSnapshot, factions FactionService, upgradeID string) UpgradePurchaseResult {
	upgrade, ok := snapshot.UpgradesByID[upgradeID]
	if !ok {
		return FailedUpgradePurchase(fmt.Sprintf("Upgrade %q does not exist.", upgradeID))
	}
	if run.Progression.HasUpgrade(upgrade.ID) {
		return FailedUpgradePurchase(fmt.Sprintf("%s is already installed.", upgrade.Name))
	}
	if upgrade.RequiredFactionID != "" &&
		!factions.MeetsMinimumStanding(*run, snapshot, upgrade.RequiredFactionID, upgrade.MinimumStanding) {
		return FailedUpgradePurchase(fmt.Sprintf("%s requires %s standing with %s.", upgrade.Name, upgrade.MinimumStanding, upgrade.RequiredFactionID))
	}
	if run.Player.Credits < upgrade.CostCredits {
		return FailedUpgradePurchase(fmt.Sprintf("You need %d credits to install %s.", upgrade.CostCredits, upgrade.Name))
	}

	run.Player.Credits -= upgrade.CostCredits
	run.Progression.PurchasedUpgradeIDs = append(run.Progression.PurchasedUpgradeIDs, upgrade.ID)
	sort.Strings(run.Progression.PurchasedUpgradeIDs)
	if upgrade.Specialization != "" {
		if run.Progression.SpecializationFlags == nil {
			run.Progression.SpecializationFlags = make(map[string]bool)
		}
		run.Progression.SpecializationFlags[string(upgrade.Specialization)] = true
	}

	for _, effect := range upgrade.Effects {
		if effect.Type == domain.UpgradeEffectCargoLimitBonus {
			run.Player.CargoLimit += effect.Value
		}
	}

	return SuccessfulUpgradePurchase(fmt.Sprintf("Installed %s for %d credits.", upgrade.Name, upgrade.CostCredits))
}

func (s UpgradeService) AdjustTravelCost(run domain.RunState, baseCost int, snapshot domain.DataSnapshot) int {
	discountPercent := 0
	for _, upgradeID := range run.Progression.PurchasedUpgradeIDs {
		upgrade, ok := snapshot.UpgradesByID[upgradeID]
		if !ok {
			continue
		}
		for _, effect := range upgrade.Effects {
			if effect.Type == domain.UpgradeEffectTravelCostDiscountPct {
				discountPercent += effect.Value
			}
		}
	}

	if discountPercent <= 0 {
		return baseCost
	}
	if discountPercent > 90 {
		discountPercent = 90
	}

	adjustedCost := baseCost * (100 - discountPercent) / 100
	if baseCost > 0 && adjustedCost < 1 {
		return 1
	}
	return adjustedCost
}

func (s UpgradeService) AdjustMissionReward(run domain.RunState, baseReward int, snapshot domain.DataSnapshot) int {
	bonusPercent := 0
	for _, upgradeID := range run.Progression.PurchasedUpgradeIDs {
		upgrade, ok := snapshot.UpgradesByID[upgradeID]
		if !ok {
			continue
		}
		for _, effect := range upgrade.Effects {
			if effect.Type == domain.UpgradeEffectMissionRewardBonusPct {
				bonusPercent += effect.Value
			}
		}
	}

	if bonusPercent <= 0 {
		return baseReward
	}

	return baseReward + (baseReward*bonusPercent)/100
}

File: src/services/UpgradeService.cs
using StarSmugglerGo.Domain;
using System.Collections.Generic;
using System.Linq;

namespace StarSmugglerGo.Services;

public sealed class UpgradeService
{
    public IReadOnlyList<ShipUpgradeDefinition> GetVisibleUpgrades(DataSnapshot data)
        => data.Upgrades
            .Where(upgrade => string.IsNullOrWhiteSpace(upgrade.RequiredFactionId))
            .OrderBy(upgrade => upgrade.CostCredits)
            .ThenBy(upgrade => upgrade.Name)
            .ToList();

    public bool CanPurchase(RunState run, ShipUpgradeDefinition upgrade)
    {
        if (run.Progression.HasUpgrade(upgrade.Id))
        {
            return false;
        }

        if (!string.IsNullOrWhiteSpace(upgrade.RequiredFactionId))
        {
            return false;
        }

        return run.Player.Credits >= upgrade.CostCredits;
    }

    public string DescribeAvailability(RunState run, ShipUpgradeDefinition upgrade)
    {
        if (run.Progression.HasUpgrade(upgrade.Id))
        {
            return "Installed";
        }

        if (!string.IsNullOrWhiteSpace(upgrade.RequiredFactionId))
        {
            return $"Locked: requires {upgrade.MinimumStanding} standing with {upgrade.RequiredFactionId}.";
        }

        if (run.Player.Credits < upgrade.CostCredits)
        {
            return $"Need {upgrade.CostCredits} credits.";
        }

        return "Ready to install.";
    }

    public UpgradePurchaseResult PurchaseUpgrade(RunState run, DataSnapshot data, string upgradeId)
    {
        if (!data.UpgradesById.TryGetValue(upgradeId, out ShipUpgradeDefinition? upgrade))
        {
            return UpgradePurchaseResult.Failure($"Upgrade '{upgradeId}' no longer exists.");
        }

        if (run.Progression.HasUpgrade(upgrade.Id))
        {
            return UpgradePurchaseResult.Failure($"{upgrade.Name} is already installed.");
        }

        if (!string.IsNullOrWhiteSpace(upgrade.RequiredFactionId))
        {
            return UpgradePurchaseResult.Failure(
                $"{upgrade.Name} requires {upgrade.MinimumStanding} standing with {upgrade.RequiredFactionId}.");
        }

        if (run.Player.Credits < upgrade.CostCredits)
        {
            return UpgradePurchaseResult.Failure($"You need {upgrade.CostCredits} credits to install {upgrade.Name}.");
        }

        run.Player.Credits -= upgrade.CostCredits;
        run.Progression.PurchasedUpgradeIds.Add(upgrade.Id);
        run.Progression.PurchasedUpgradeIds.Sort(System.StringComparer.Ordinal);

        if (upgrade.Specialization is ShipSpecialization specialization)
        {
            run.Progression.SpecializationFlags[specialization.ToString()] = true;
        }

        foreach (UpgradeEffectDefinition effect in upgrade.Effects)
        {
            if (effect.Type == UpgradeEffectType.CargoLimitBonus)
            {
                run.Player.CargoLimit += effect.Value;
            }
        }

        return UpgradePurchaseResult.Success($"Installed {upgrade.Name} for {upgrade.CostCredits} credits.");
    }

    public int AdjustTravelCost(RunState run, DataSnapshot data, int baseCost)
    {
        int discountPercent = 0;
        foreach (string upgradeId in run.Progression.PurchasedUpgradeIds)
        {
            if (!data.UpgradesById.TryGetValue(upgradeId, out ShipUpgradeDefinition? upgrade))
            {
                continue;
            }

            foreach (UpgradeEffectDefinition effect in upgrade.Effects)
            {
                if (effect.Type == UpgradeEffectType.TravelCostDiscountPercent)
                {
                    discountPercent += effect.Value;
                }
            }
        }

        if (discountPercent <= 0)
        {
            return baseCost;
        }

        discountPercent = System.Math.Min(discountPercent, 90);
        int adjustedCost = baseCost * (100 - discountPercent) / 100;
        return baseCost > 0 ? System.Math.Max(1, adjustedCost) : adjustedCost;
    }

    public int GetCheapestTravelCostFromPort(RunState run, DataSnapshot data, TravelService travelService, PortDefinition origin)
    {
        int baseCost = travelService.GetCheapestTravelCostFromPort(origin, data.Ports);
        return AdjustTravelCost(run, data, baseCost);
    }

    public string SummarizeEffects(ShipUpgradeDefinition upgrade)
    {
        return string.Join(", ", upgrade.Effects.Select(effect => effect.Type switch
        {
            UpgradeEffectType.CargoLimitBonus => $"+{effect.Value} cargo",
            UpgradeEffectType.TravelCostDiscountPercent => $"-{effect.Value}% travel cost",
            UpgradeEffectType.MissionRewardBonusPercent => $"+{effect.Value}% mission rewards",
            _ => $"{effect.Type} {effect.Value}",
        }));
    }
}
